### PR TITLE
Add RPC PoET

### PR DESCRIPTION
**Pull request by moshababo**
Added implementation of RPC PoET client/server to be used in the NIPST construction flow (mostly for testing purposes).

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,29 +1,27 @@
 module github.com/spacemeshos/go-spacemesh
 
 require (
-	github.com/BurntSushi/toml v0.3.1 // indirect
 	github.com/VividCortex/gohistogram v1.0.0 // indirect
-	github.com/btcsuite/btcd v0.0.0-20181130015935-7d2daa5bfef2
-	github.com/btcsuite/btcutil v0.0.0-20180706230648-ab6388e0c60a
+	github.com/btcsuite/btcd v0.0.0-20190213025234-306aecffea32
```
Comment (antonlerner) [inline go.mod:5]: have a quick look at  these dependencies, just make sure no redundant dependancies were added

File: nipst/nipst_test.go [hunk 2]
```diff
@@ -58,10 +61,55 @@ func TestNIPSTBuilder(t *testing.T) {
...
-		t.Fail()
+	case <-time.After(5 * time.Second):
+		assert.Fail("timeout")
 		return
 	}
 
 	nb.Stop()
 }
+
+func TestNIPSTBuilderWithRPCClients(t *testing.T) {
```
Comment (antonlerner) [inline nipst/nipst_test.go:73]: need to add a test that actually validates the created nipst

File: nipst/poet.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+	return p, nil
+}
+
+func (c *RPCPoetHarness) cleanUp() error {
+	return c.h.TearDown()
+}
+
+func wireLabelsToNative(in []*api.Labels) (native [shared.T]shared.Labels) {
+	for i, inLabels := range in {
```
Comment (antonlerner) [inline nipst/poet.go]: checjk that len(in) == shared.T

**Review by antonlerner — CHANGES_REQUESTED** [review]
please fix fmt test as well

File: go.mod (hunk 1 above)
Comment (moshababo) [reply go.mod]: Since this dependency was already existing, let's wait with attempting to remove it.

File: nipst/poet.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+		return nil, err
+	}
+
+	return &membershipProof{merkleProof: res.MerkleProof}, nil
+}
+
+func (c *RPCPoetClient) subscribeProof(r *poetRound,
+	timeout time.Duration) (*poetProof, error) {
+
+	ctx, _ := context.WithTimeout(context.Background(), timeout)
```
Comment (antonlerner) [inline nipst/poet.go]: check error

File: nipst/poet.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+		return nil, err
+	}
+
+	return &poetRound{id: int(res.RoundId)}, nil
+}
+
+func (c *RPCPoetClient) subscribeMembershipProof(r *poetRound,
+	challenge common.Hash, timeout time.Duration) (*membershipProof, error) {
+
+	ctx, _ := context.WithTimeout(context.Background(), timeout)
```
Comment (antonlerner) [inline nipst/poet.go]: dont need to defer the cancel funciton?

**Review by antonlerner — APPROVED** [review]

File: nipst/poet.go (hunk 4 above)
Comment (moshababo) [reply nipst/poet.go]: It's not an error here, just the cancellation function which is not necessary.

File: nipst/poet.go (hunk 5 above)
Comment (moshababo) [reply nipst/poet.go]: Right, I'll add that.

File: nipst/nipst_test.go (hunk 2 above)
Comment (moshababo) [reply nipst/nipst_test.go:73]: I added assertion to the nipst validity, which is merely a mock now. Will rewrite these tests after the upcoming changes.

File: go.mod (hunk 1 above)
Comment (moshababo) [reply go.mod:5]: Since they were already existing, i'll wait with the attempt to remove them.

**Pull request merged**